File: plugins/android/project/FirebaseCrashlytics/src/main/java/eu/novuloj/firebasecrashlytics/FirebaseCrashlytics.kt
package eu.novuloj.firebasecrashlytics

import androidx.annotation.NonNull
import com.google.firebase.crashlytics.FirebaseCrashlytics
import com.google.firebase.crashlytics.ktx.crashlytics
import com.google.firebase.ktx.Firebase
import org.godotengine.godot.Godot
import org.godotengine.godot.plugin.GodotPlugin
import org.godotengine.godot.plugin.UsedByGodot

class FirebaseCrashlytics(godot: Godot) : GodotPlugin(godot) {
    private val pluginName = "FirebaseCrashlytics"
    private lateinit var crashlytics: FirebaseCrashlytics

    @NonNull
    override fun getPluginName(): String = pluginName

    @UsedByGodot
    fun init() {
        crashlytics = Firebase.crashlytics
    }

    @UsedByGodot
    fun checkForUnsentReports(): Boolean = false // TODO

    @UsedByGodot
    fun deleteUnsentReports() = crashlytics.deleteUnsentReports()

    @UsedByGodot
    fun didCrashOnPreviousExecution(): Boolean = crashlytics.didCrashOnPreviousExecution()

    @UsedByGodot
    fun log(message: String) = crashlytics.log(message)

    @UsedByGodot
    fun recordException(message: String) {
        val throwable: Throwable = Throwable(message)
        crashlytics.recordException(throwable);
    }

    @UsedByGodot
    fun sendUnsentReports() = crashlytics.sendUnsentReports()

    @UsedByGodot
    fun setCrashlyticsCollectionEnabled(enabled: Boolean) =
        crashlytics.setCrashlyticsCollectionEnabled(enabled)

    @UsedByGodot
    fun setCustomKeyB(key: String, value: Boolean) = crashlytics.setCustomKey(key, value)

    @UsedByGodot
    fun setCustomKeyF(key: String, value: Float) = crashlytics.setCustomKey(key, value)

    @UsedByGodot
    fun setCustomKeyI(key: String, value: Int) = crashlytics.setCustomKey(key, value)

    @UsedByGodot
    fun setCustomKeyS(key: String, value: String) = crashlytics.setCustomKey(key, value)

    @UsedByGodot
    fun setUserId(identifier: String) = crashlytics.setUserId(identifier)

    @UsedByGodot
    fun crash(): Boolean = throw RuntimeException("From $pluginName")
}

File: addons/lavos/plugins/google/firebase/crashlytics/FirebaseCrashlytics.cs
using Godot;
using Lavos.Services.Crash;
using System;

namespace Lavos.Plugins.Google.Firebase.Crashlytics;

sealed class FirebaseCrashlytics : ICrashService
{
    const string PluginName = "FirebaseCrashlytics";
    readonly LavosPlugin Plugin;

    public FirebaseCrashlytics()
    {
        Assert.IsTrue(Engine.HasSingleton(PluginName), $"Missing plugin {PluginName}");
        Plugin = new LavosPlugin(Engine.GetSingleton(PluginName));
    }

    public void Initialise()
    {
        Plugin.CallVoid("init");
    }

    #region ICrashService

    public void EnableCollection(bool enable)
    {
        Plugin.CallVoid("setCrashlyticsCollectionEnabled", enable);
    }

    public bool CheckForUnsentReports()
    {
        return Plugin.CallBool("checkForUnsentReports");
    }

    public void SendUnsentReports()
    {
        Plugin.CallVoid("sendUnsentReports");
    }

    public void DeleteUnsentReports()
    {
        Plugin.CallVoid("deleteUnsentReports");
    }

    public bool DidCrashOnPreviousExecution()
    {
        return Plugin.CallBool("didCrashOnPreviousExecution");
    }

    public void Log(string message)
    {
        Plugin.CallVoid("log", message);
    }

    public void LogException(Exception e)
    {
        var exception = $"{e.GetType()} | {e.Message}";
        Plugin.CallVoid("recordException", exception);
    }

    public void SetUserId(string id)
    {
        Plugin.CallVoid("setUserId", id);
    }

    public void SetCustomKey(string key, Variant value)
    {
        var method = string.Empty;
        switch (value.VariantType)
        {
            case Variant.Type.Int: Plugin.CallVoid("setCustomKeyI", key, value); return;
            case Variant.Type.Float: Plugin.CallVoid("setCustomKeyF", key, value); return;
            case Variant.Type.Bool: Plugin.CallVoid("setCustomKeyB", key, value); return;
            case Variant.Type.String: Plugin.CallVoid("setCustomKeyS", key, value); return;
            default:
                {
                    Plugin.CallVoid("setCustomKeyS", key, "TypeNotSupported");
                    Console.Log.Error(nameof(FirebaseCrashlytics), "Unhandled custom type");
                    return;
                }
        }
    }

    public void NativeCrash()
    {
#if DEBUG
        Plugin.CallVoid("crash");
#endif
    }

    #endregion ICrashService
}